src/interval_sets.jl

```diff
@@ -1,29 +1,76 @@
+1-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(3, 5)
+ 
+julia> Array(symdiff(IntervalSet(1..5), IntervalSet(3..8)))
+2-element Vector{Interval{Int64}}:
+ Interval{Int64, Closed, Open}(1, 3)
+ Interval{Int64, Open, Closed}(5, 8)
+
+julia> Array(union(IntervalSet([1..2, 2..5]), IntervalSet(6..7)))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 5)
+ Interval{Int64, Closed, Closed}(6, 7)
+
+julia> Array(union(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(1, 10)
+ Interval{Int64, Closed, Closed}(12, 14)
+
+julia> Array(intersect(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64, Closed, Closed}}:
+ Interval{Int64, Closed, Closed}(4, 5)
+ Interval{Int64, Closed, Closed}(8, 9)
+
+julia> Array(setdiff(IntervalSet([1..5, 8..10]), IntervalSet([4..9, 12..14])))
+2-element Vector{Interval{Int64}}:
+ Interval{Int64, Closed, Open}(1, 4)
+ Interval{Int64, Open, Closed}(9, 10)
+```
 """
-struct IntervalSet{T<:AbstractInterval} <: AbstractSet{T}
-    # TODO: Use Dict internally once union!(Vector{AbstractInterval}) is fully deprecated
-    items::Vector{T}
+struct IntervalSet{I <: AbstractInterval, A <: AbstractVector{I}}
+    items::A
 end
 
-IntervalSet(v::AbstractVector) = IntervalSet{eltype(v)}(v)
-IntervalSet(interval::T) where T <: AbstractInterval = IntervalSet{T}([interval])
+IntervalSet{T}(intervals) where T <: AbstractInterval = IntervalSet{T, Vector{T}}(intervals)
+IntervalSet(interval::T) where T <: AbstractInterval = IntervalSet{T, Vector{T}}([interval])
 IntervalSet(interval::IntervalSet) = interval
 IntervalSet(itr) = IntervalSet{eltype(itr)}(collect(itr))
-IntervalSet() = IntervalSet{AbstractInterval}(AbstractInterval[])
+IntervalSet() = IntervalSet(AbstractInterval[])
 
-Base.copy(intervals::IntervalSet{T}) where T = IntervalSet{T}(copy(intervals.items))
+Base.copy(intervals::IntervalSet{T, A}) where {T, A} = IntervalSet{T, A}(copy(intervals.items))
 Base.length(intervals::IntervalSet) = length(intervals.items)
 Base.iterate(intervals::IntervalSet, args...) = iterate(intervals.items, args...)
 Base.eltype(::IntervalSet{T}) where T = T
-Base.:(==)(a::IntervalSet, b::IntervalSet) = issetequal(a, b)
+Base.:(==)(a::IntervalSet, b::IntervalSet) = a.items == b.items
+Base.isequal(a::IntervalSet, b::IntervalSet) = isequal(a, b)
+Base.Array(intervals::IntervalSet) = intervals.items
 
-const AbstractIntervals = Union{AbstractInterval, IntervalSet}
+# currently (to avoid breaking changes) new methods for `Base`
+# accept `IntervalSet` objects and Interval singletons.
+const AbstractIntervalSets = Union{AbstractInterval, IntervalSet}
+# internal methods need to operate on both sets and arrays of intervals
+const AbstractIntervals = Union{AbstractIntervalSets, AbstractArray{<:AbstractInterval}}
```

Comment:
I'd like to move away from dispatching on vectors of intervals. I think we should make an explicit dispatch when we want to do that rather than having this union type floating around which risks introducing more breaking changes.